packages/flutter/lib/src/widgets/framework.dart

```diff
@@ -1126,8 +1126,43 @@ abstract class State<T extends StatefulWidget> with Diagnosticable {
   /// ```
   ///
   /// It is an error to call this method after the framework calls [dispose].
-  /// You can determine whether it is legal to call this method by checking
+  /// You can determine whether it is legal to call [setState] by checking
   /// whether the [mounted] property is true.
+  ///
+  /// # Performance considerations
+  ///
+  /// Calls to [setState] cause a rebuild of the current widget and,
+  /// potentially, all of its children. Generally, [setState] calls should be
+  /// made on widgets with small subtrees in order to rebuild the least amount
+  /// of children.
+  ///
+  /// If [setState] was frequently called by a ticker or a listener for example,
+  /// move the state into a separated widget. In this way, rebuilds would be
+  /// limited to a single widget rather than multiple ones that might not even
+  /// need to be updated. This technique of pushing the state into leaf widgets
+  /// is a good practice.
+  ///
+  /// In the case where the state cannot be pushed into a leaf widget, there are
+  /// some alternatives:
+  ///
+  ///   1. Using a [ValueNotifier] to hold the state of a single value and a
+  ///      [ValueListenableBuilder] to listen for changes.
+  ///
+  ///   2. Using a [Listenable] (such as [ChangeNotifier] for example) to hold
+  ///      the state and an [AnimatedBuilder] to rebuild the children.
+  ///
+  /// While [setState] could rebuild the entire subtree, builder widgets like
+  /// [AnimatedBuilder] and [ValueListenableBuilder] only rebuild their content
+  /// leaving the enclosing widget and its children untouched. This is efficient
+  /// when the state has to be located up in the tree and only certain widgets
+  /// need to be updated.
+  ///
+  /// See also:
+  ///
+  ///  * [AnimatedBuilder], in the "Improve rebuilds performance using
+  ///    AnimatedBuilder" section to see how to efficiently manage the state
+  ///    using a [Listenable].
+  ///  * [StatefulWidget], in the "Performance considerations" section.
```

Comment:
Adding this link is an excellent idea! It describes in a neutral way what can be done to minimize the impact of rebuilding a stateful widget and people should be aware of that.

However, as has been said before on this PR, we need to be very careful to not scare people away from using `setState`. It is after all a central building block of Flutter apps.